tfhe/src/core_crypto/experimental/entities/lwe_shrinking_keyswitch_key.rs

```diff
@@ -0,0 +1,328 @@
+            self.decomposition_level_count(),
+            self.input_key_lwe_dimension(),
+            self.output_lwe_size(),
+            self.ciphertext_modulus(),
+        )
+    }
+}
+
+impl<Scalar: UnsignedInteger, C: ContainerMut<Element = Scalar>> LweShrinkingKeyswitchKey<C> {
+    pub fn as_mut_lwe_keyswitch_key(&mut self) -> LweKeyswitchKey<&'_ mut [Scalar]> {
+        self.lwe_ksk.as_mut_view()
+    }
+
+    /// Mutable variant of [`LweShrinkingKeyswitchKey::as_view`].
+    pub fn as_mut_view(&mut self) -> LweShrinkingKeyswitchKey<&'_ mut [Scalar]> {
+        let decomposition_base_log = self.decomposition_base_log();
+        let decomposition_level_count = self.decomposition_level_count();
+        let input_key_lwe_dimension = self.input_key_lwe_dimension();
+        let output_lwe_size = self.output_lwe_size();
+        let ciphertext_modulus = self.ciphertext_modulus();
+        LweShrinkingKeyswitchKey::from_container(
+            self.as_mut(),
+            decomposition_base_log,
+            decomposition_level_count,
+            input_key_lwe_dimension,
+            output_lwe_size,
+            ciphertext_modulus,
+        )
+    }
+}
+
+/// An [`LweShrinkingKeyswitchKey`] owning the memory for its own storage.
+pub type LweShrinkingKeyswitchKeyOwned<Scalar> = LweShrinkingKeyswitchKey<Vec<Scalar>>;
+/// A [`LweShrinkingKeyswitchKey`] immutably borrowing memory for its own storage.
+pub type LweShrinkingKeyswitchKeyView<'data, Scalar> = LweShrinkingKeyswitchKey<&'data [Scalar]>;
+/// A [`LweShrinkingKeyswitchKey`] mutably borrowing memory for its own storage.
+pub type LweShrinkingKeyswitchKeyMutView<'data, Scalar> =
+    LweShrinkingKeyswitchKey<&'data mut [Scalar]>;
+
+impl<Scalar: UnsignedInteger> LweShrinkingKeyswitchKeyOwned<Scalar> {
+    /// Allocate memory and create a new owned [`LweShrinkingKeyswitchKey`].
+    ///
+    /// # Note
+    ///
+    /// This function allocates a vector of the appropriate size and wraps it in the appropriate
+    /// type. If you want to generate an LWE shrinking keysiwtch key you need to use
+    /// [`crate::core_crypto::experimental::algorithms::generate_lwe_shrinking_keyswitch_key`] using
+    /// this key as output.
+    ///
+    /// See [`LweShrinkingKeyswitchKey::from_container`] for usage.
+    pub fn new(
+        fill_with: Scalar,
+        decomp_base_log: DecompositionBaseLog,
+        decomp_level_count: DecompositionLevelCount,
+        input_key_lwe_dimension: LweDimension,
+        output_key_lwe_dimension: LweDimension,
+        shared_randomness_coef_count: LweSecretKeySharedCoefCount,
+        ciphertext_modulus: CiphertextModulus<Scalar>,
+    ) -> Self {
+        assert!(
+            output_key_lwe_dimension.0 <= input_key_lwe_dimension.0,
+            "The output LweDimension ({output_key_lwe_dimension:?}) \
+            must be smaller than the input LweDimension ({input_key_lwe_dimension:?}) \
+            for an LweShrinkingKeyswitchKey."
+        );
```

Comment:
Is this necessary?

Reply:
probably not

Reply:
we will keep it as discussed as the other way around for shared keys would mean filling the end of a mask with 0s instead of actually keyswitching